File: searches/binary_search.js
/**
 * Binary Search implementation in JavaScript
 * @param {Array} arr - Sorted array
 * @param {number} target - Value to search for
 * @returns {number} - Index of target or -1
 */
function binarySearch(arr, target) {
    let left = 0;
    let right = arr.length - 1;

    while (left <= right) {
        let mid = Math.floor(left + (right - left) / 2);

        if (arr[mid] === target) {
            return mid;
        }
        if (arr[mid] < target) {
            left = mid + 1;
        } else {
            right = mid - 1;
        }
    }
    return -1;
}

// Example usage
const arr = [2, 3, 4, 10, 40];
const target = 10;
const result = binarySearch(arr, target);

console.log(`Element found at index: ${result}`);

File: searches/binary_search.go
package main

import "fmt"

// BinarySearch returns the index of target in arr, or -1 if not found
func BinarySearch(arr []int, target int) int {
	left := 0
	right := len(arr) - 1

	for left <= right {
		mid := left + (right-left)/2

		if arr[mid] == target {
			return mid
		}
		if arr[mid] < target {
			left = mid + 1
		} else {
			right = mid - 1
		}
	}
	return -1
}

func main() {
	arr := []int{2, 3, 4, 10, 40}
	target := 10
	result := BinarySearch(arr, target)
	
	if result != -1 {
		fmt.Printf("Element found at index %d\n", result)
	} else {
		fmt.Println("Element not found")
	}
}
